packages/plot/src/utils.rs

```diff
@@ -0,0 +1,402 @@
+            Ok(Self {
+                from,
+                to,
+                step,
+                curr: from,
+            })
+        }
+    }
+
+    pub fn peek(&self) -> Option<f64> {
+        if self.curr <= self.to {
+            Some(self.curr)
+        } else {
+            None
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        ((self.to - self.from) / self.step) as usize
+    }
+}
+
+// Used for build_cartesian_2d
+impl TryFrom<FloatIterator> for Range<f64> {
+    type Error = String;
+
+    fn try_from(value: FloatIterator) -> Result<Self, Self::Error> {
+        if value.step != 1.0 {
+            Err(String::from(
+                "Can not convert to Range<f64> when step is not 1.0.",
+            ))
+        } else {
+            Ok(value.from..value.to)
+        }
+    }
+}
+
+// Should be relatively easy to extend this to three axes if needed
+pub struct RangeManager {
+    user_range_x: (f64, f64),
+    user_range_y: (f64, f64),
+    points: f64,
+}
+
+impl RangeManager {
+    pub fn new(points: u64, x_fr: f64, x_to: f64, y_fr: f64, y_to: f64) -> Self {
+        Self {
+            user_range_x: (x_fr, x_to),
+            user_range_y: (y_fr, y_to),
+            points: points as f64,
+        }
+    }
+
+    pub fn get_user_range_endpoints(&self, axis: char) -> Result<(f64, f64), String> {
+        match axis {
+            'x' => Ok((self.user_range_x.0, self.user_range_x.1)),
+            'y' => Ok((self.user_range_y.0, self.user_range_y.1)),
+            _ => Err(String::from("Invalid axis")),
+        }
+    }
+
+    pub fn get_user_range(&self, axis: char) -> Result<FloatIterator, String> {
+        let (fr, to) = self.get_user_range_endpoints(axis)?;
+        FloatIterator::new(fr, to, 1.0)
+    }
+
+    pub fn get_point_range(&self, axis: char) -> Result<FloatIterator, String> {
+        let (fr, to) = self.get_user_range_endpoints(axis)?;
+        // Subtracting with f64 epsilon because otherwise floating point arithmetic may result in
+        // last value being just above "to", example 4+1e^-16 is left out of (-4 to 4)
+        let step = (to - fr) / self.points - f64::EPSILON;
+        FloatIterator::new(fr, to, step)
+    }
+
+    // Returns 0 if y is inside the boundaries, otherwise the distance to the closest boundary
+    pub fn y_margin(&self, y: f64) -> f64 {
+        let (min, max) = self.user_range_y;
+        if y > max {
+            y - max
+        } else if y < min {
+            y - min
+        } else {
+            0.0
+        }
+    }
+}
+
+pub fn print_svg_html(svg: String, ctx: &PlotContext) {
+    let encoded: String = general_purpose::STANDARD_NO_PAD.encode(&svg);
+    let src = format!("data:image/svg+xml;base64,{encoded}");
+    let percentage = (ctx.svg_info.width * 100.0).round() as i32;
+    let style = format!("style=\"width:{percentage}%\"");
+    let img_str = format!("<img src=\"{src}\" {style} ");
+
+    let mut v = vec![];
+    v.push(raw!("<figure>\n"));
```

Comment:
Since we now (as far as I know?) just support string literals for raw modules we maybe be able to remove the use of the `raw!` macro in favour of using the string : )